README.md
RestrictedTextField v1.4
========================

See LICENSE for this software's licensing terms.

RestrictedTextField is a jQuery plugin which uses regular expressions to validate and control input to HTML text fields.  RestrictedTextField has 38 built-in types and also allows you to define your own types.  Validation is performed on keystroke, paste and blur.

You might not need this plugin for your purposes - the HTML5 `pattern` attribute provides most of the same functionality.  If, however, you must support browsers that don't support `pattern`, or if you want to make use of RestrictedTextField's additional features, then perhaps it might be a good fit for you.

The biggest difference between this plugin and `pattern` is that RestrictedTextField allows you to (optionally) suppress invalid keystrokes in real time, rather than allow them into the field to fail validation later.  If you do want to allow invalid characters in, however, you have that option.

If you're only interested in `pattern`-based validation, you might still find RestrictedTextField useful.  It provides the ability to disable its own validation implementation and use `pattern` exclusively.  In this scenario, you'd just be using RestrictedTextField as a repository of pre-written regular expressions.  Although you could just copy the regular expressions from the source code, you might find it easier on code maintenance to not be dropping multiple copies of the same regular expressions into your project.


## Features

* Discard invalid keystrokes or catch a validation failure event to handle it as you wish
* Has 38 built-in types
* Extensible:  define your own types
* Supports (but does not rely on) the HTML 5 `pattern` attribute - can be enabled during initialization
* Money types automatically format on blur to end in a decimal point and two digits


## Built-in Types

1.  Alpha strings
2.  Uppercase alpha strings
3.  Lowercase alpha strings
4.  Alpha strings plus space
5.  Uppercase alpha strings plus space
6.  Lowercase alpha strings plus space
7.  Alphanumeric strings
8.  Uppercase alphanumeric strings
9.  Lowercase alphanumeric strings
10.  Alphanumeric strings plus space
11.  Uppercase alphanumeric strings plus space
12.  Lowercase alphanumeric strings plus space
13.  Integers
14.  Positive integers
15.  Negative integers
16.  Strict integers
17.  Strict positive integers
18.  Strict negative integers
19.  Floats
20.  Positive floats
21.  Negative floats
22.  Strict floats
23.  Strict positive floats
24.  Strict negative floats
25.  Money
26.  Positive money
27.  Negative money
28.  Accounting money notation (negatives expressed by wrapping the value in parentheses)
29.  Negative accounting money
30.  American Express
31.  VISA 
32.  MasterCard
33.  Discover
34.  Credit Card (combines all of the individual credit card types above)
35.  [Luhn](https://en.wikipedia.org/wiki/Luhn_algorithm) numbers - note:  if used with `pattern` mode enabled, successful validation is essentially meaningless.
36.  US zip code (5 digits)
37.  US zip code suffix (1-4 digits)
38.  US zip code (5 digits with optional suffix)


## What's the difference between integer/strict integer, float/strict float?

### Integer types:

* Zero can be negative

### Strict integer types:

* Zero cannot be negative

### Float types:

* Integers are valid
* Zero is valid (0, 0.0, .0, 0000.000, etc.)
* Zero can be negative

### Strict float types:

* Values must be floating-point
* Zero is valid (0.0, .0, 0000.000, etc.)
* Zero cannot be negative


## Credit Card types

Each credit card type enforces prefixes and length as specified by https://en.wikipedia.org/wiki/Payment_card_number in December, 2016.  [Luhn validation](https://en.wikipedia.org/wiki/Luhn_algorithm) is also performed.

If you prefer not to use strictly-enforced credit card types (is Wikipedia correct, and is RestrictedTextField up-to-date?), you can use the Luhn type instead.  It performs basic mathematical validation, but that will only tell you that a credit card number is invalid, not that it IS valid.  Full validation would then be shifted to your payment card processor in the form of a rejected transaction -- or, if they provide an API for validation, you could use that, too.


## Configuration

| Property | Description   | Data Type | Valid Values         | Default Value |
| -------- | --------------|---------- |----------------------|---------------|
| `type`   | Text field type.  This is a required setting. | string | alpha, upperAlpha, lowerAlpha, alphaSpace, upperAlphaSpace, lowerAlphaSpace, alphanumeric, upperAlphanumeric, lowerAlphanumeric, alphanumericSpace, upperAlphanumericSpace, lowerAlphanumericSpace, int, positiveInt, negativeInt, strictInt, strictPositiveInt, strictNegativeInt, float, positiveFloat, negativeFloat, strictFloat, strictPositiveFloat, strictNegativeFloat, money, positiveMoney, negativeMoney, accountingMoney, negativeAccountingMoney, americanExpress, visa, masterCard, discover, creditCard, luhnNumber, usZip, usZip5, usZipSuffix | null |
| `preventInvalidInput` | When enabled, invalid keystrokes are ignored (the value of the text field is not updated).  When disabled, invalid keystrokes are not ignored. | boolean | true/false | true |
| `logger` | An optional callback function for logging.  If you want to enable logging, provide a function and then do whatever you wish with the message. | function | A function accepting the log message as a string argument | undefined |
| `usePatternAttr` | When enabled, sets the HTML5 `pattern` attribute on the input field.  Check [caniuse.com](http://caniuse.com/#search=pattern) for browser support.  When pattern mode is enabled, RestrictedTextField does not fire its own validation events, nor does it perform Luhn validation on credit card types or the Luhn type.  Pattern mode is meant to be pure HTML5 behavior.  The only part of RestrictedTextField functionality that remains active is the automatic formatting of money types.| boolean | true/false | false |


## Events

These events are fired based on the state of the text field, but only if `usePatternAttr` is set to false.

| Event name        | Description                                                     |
| ------------------| ----------------------------------------------------------------|
| inputIgnored      | Fires when an invalid keystroke is ignored.  `preventInvalidInput` must be enabled for this event to fire. |
| validationFailed  | If `preventInvalidInput` is disabled, fires when an invalid keystroke is made.  Also fires if validation performed on blur fails. |
| validationSuccess | If `preventInvalidInput` is disabled, fires when the user removes invalid data.  Also fires if validation performed on blur succeeds. |


#### Example


```html
<input type="text" id="field"/>
```
```javascript
$( "#field" ).restrictedTextField( {
   type : "alpha"
} );
```

See `demo/demo.html` for detailed examples.


# Desktop Browser Compatibility

Many other versions of Firefox and Chrome, whether older or newer than what's listed here, are certainly compatible.  Browsers not listed here might work fine, but have not been tested.  If jQuery is supported by the browser, RestrictedTextField should also work.  If you'd like to contribute compatibility information, send it to the e-mail address listed on my GitHub profile and I'll post it here.

* Firefox 49.0.2
* Pale Moon 26.4.1
* Chrome 54.0.2840.99 m
* Edge
* IE 11
* IE 10
* IE 9
* IE 8:  INCOMPATIBLE


# Mobile Browser Compatibility (Android)

Many other versions of the listed browsers, whether older or newer that what's listed here, are certainly compatible.  Browsers not listed here might work fine, but have not been tested.  If jQuery is supported by the browser, RestrictedTextField should also work.  If you'd like to contribute compatibility information, send it to the e-mail address listed on my GitHub profile and I'll post it here.

* Firefox 50.1.0
* Pale Moon 25.9.6
* Chrome 55.0.2883.91
* Opera 4.3 / 41.1.2246.111645 - listed in the Play store as 4.3; reports as 41.1.2246.111645 from within Opera


## Donations

https://paypal.me/KurtisLoVerde/10

Thank you for your support!


CHANGELOG.md
# Release 1.4 (February 10, 2018)

* Added support for the HTML5 `pattern` attribute.  Check [caniuse.com](http://caniuse.com/#search=pattern) for browser support.
* Added the ability to save the QUnit test report to a file


# Release 1.3 (January 3, 2017)

* Added zip code types


# Release 1.2 (December 28, 2016)

* Now confirmed to work on mobile!  See README.md for a list of tested browsers.
* Change to event firing:  validation success & validation failure events were firing pre-blur with `preventInvalidInput` enabled.  This was contrary to the documentation, which stated that these events do not fire in that scenario.  The documentation reflected the intended behavior; the events no longer fire in this scenario.
* Added credit card types (AMEX, VISA, MasterCard, Discover), plus a credit card type encompassing each of these types
* Added a [Luhn](https://en.wikipedia.org/wiki/Luhn_algorithm) type
* Updated and redesigned the demo page
* The unit tests are no longer implemented with Java/Selenium/JUnit; they've been re-implemented in JavaScript/QUnit.  This dramatically improves test performance and greatly simplifies running tests (just open the HTML file).


# Release 1.1.1 (November 28, 2016)

* The demo page wasn't working due to an incorrect path to jquery.js.  This has been fixed.  Although the source code, including the JavaScript files themselves, have been updated to say 1.1.1 in their internal code comments, there is no difference in these files from release 1.1.  Only demo.html has been changed.


# Release 1.1 (November 15, 2016)

* Fixes for money types
* All money types automatically format on blur:  values will be updated, if necessary, to always end in a period followed by two digits
* Fixed issue with the int types which considered a negative sign by itself to be a valid value
* Relaxed the int and float types/subtypes, and added strict versions of each.  The relaxed versions accept leading/trailing zeros and negative zero; the strict versions do not.
* Added a logging callback function as a configurable option
* Throws an exception if invoked on something other than an input element
* Added Selenium unit tests


# Release 1.0.1 (April 2, 2016)

* Minor under-the-hood improvement.  No change to functionality.


## Release 1.0 (April 2, 2016)

* First release


Gruntfile.js
module.exports = function(grunt) {

   grunt.initConfig( {
      //properties: {
      //   props: "grunt.properties"
      //},

      pkg: grunt.file.readJSON('package.json'),

      uglify: {
         options: {
            banner: "/*! <%= pkg.name %> <%= grunt.template.today('yyyy-mm-dd') %> */\n"
         },

         build: {
            src  : "src/<%= pkg.name %>.js",
            dest : "build/<%= pkg.name %>.min.js"
         }
      },

      // Gradle builds and runs the Selenium client (JUnit tests)
      exec: {
         "gradle-test": {
            cmd: function() {
               return "cd SeleniumTester && gradle clean test --stacktrace";
            }
         }
      },

      // The Selenium server retrieves the test page via HTTP
      "http-server": {         
         "dev": {
            root: ".",
            port: 8000,
            host: "127.0.0.1",
            cache: 0,
            showDir: true,
            autoIndex: true,
            ext: "html",
            runInBackground: true,
            openBrowser: false
         },

         // This configuration is not used by the build process.  It keeps the server running so that you can run SeleniumTester in a debugger.
         "debug": {
            root: ".",
            port: 8000,
            host: "127.0.0.1",
            cache: 0,
            showDir: true,
            autoIndex: true,
            ext: "html",
            runInBackground: false,
            openBrowser: false            
         }
      },

      // The Selenium server loads the test page and automates the browser per the requests from the Selenium client.
      // If you change the Selenium version here, be sure to update build.gradle in the SeleniumTester project.
      "start-selenium-server": {
         dev: {
            options: {
               autostop: false,
               downloadUrl: "https://selenium-release.storage.googleapis.com/3.0/selenium-server-standalone-3.0.1.jar",
               serverOptions: {},
               systemProperties: {}
            }
         }
      },

      "stop-selenium-server": {
         dev: {}
      }
   } );

   grunt.loadNpmTasks( "grunt-exec" );
   grunt.loadNpmTasks( "grunt-http-server" );
   grunt.loadNpmTasks( "grunt-selenium-server" );
   //grunt.loadNpmTasks( "grunt-properties-reader" );
   grunt.loadNpmTasks( "grunt-contrib-uglify" );

   grunt.registerTask( "test", "Run unit tests", function() {
      grunt.task.run( "http-server:dev", "start-selenium-server:dev", "exec:gradle-test", "stop-selenium-server:dev" );
   } );
};


src/jquery.restrictedtextfield.js
/*
 * RestrictedTextField v1.4
 * https://www.github.com/kloverde/jquery-RestrictedTextField
 *
 * Copyright (c) 2016, Kurtis LoVerde
 * All rights reserved.
 *
 * Donations:  https://paypal.me/KurtisLoVerde/10
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *     1. Redistributions of source code must retain the above copyright
 *        notice, this list of conditions and the following disclaimer.
 *     2. Redistributions in binary form must reproduce the above copyright
 *        notice, this list of conditions and the following disclaimer in the
 *        documentation and/or other materials provided with the distribution.
 *     3. Neither the name of the copyright holder nor the names of its
 *        contributors may be used to endorse or promote products derived from
 *        this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

(function( $ ) {
   "use strict";

   /**
    * Developers:  Instantiate this to add your own types
    */
   $.fn.RestrictedTextFieldConfig = function() {
      var dest = $.fn.RestrictedTextFieldConfig.customTypes = $.fn.RestrictedTextFieldConfig.customTypes || [];

      /**
       * Adds a custom type
       *
       * @param id           - The identifier you will use to access the new type
       *
       * @param fullRegex    - RegExp object.  This regular expression describes what a valid value looks like when the user is finished
       *                       entering data, NOT what valid input looks like as it's being entered.  If your type is complex, like a
       *                       phone number, for example, you need to provide a value for partialRegex to ensure that the user will be
       *                       able to continue to type into the field.  If partialRegex is not defined or is null, the text field will
       *                       be validated against this fullRegex on every keystroke.  Notwithstanding any of the foregoing, this
       *                       regular expression is also used to validate the text field on blur.
       *
       * @param partialRegex - RegExp object.  This regular expression describes what a valid value looks like while the user is entering
       *                       it, NOT what valid input looks like after the user has finished.  If you provide a value for this, this
       *                       regular expression will be used to validate the text field on every keystroke.  This parameter may be
       *                       null, but if your type is complex, you will find it necessary to provide a value.
       */
      this.addType = function( id, fullRegex, partialRegex ) {
         _addType( dest, id, fullRegex, partialRegex );
      }
   };

   function _addType( destination, id, fullRegex, partialRegex ) {
      if( id == undefined || id == null ) {
         throw "id is undefined";
      }

      if( !isClassName(id, "String") ) {
         throw "id must be a string";
      }

      if( id.length < 1 ) {
         throw "id is empty";
      }

      if( fullRegex == undefined || fullRegex == null ) {
         throw "fullRegex is undefined";
      }

      if( !isClassName(fullRegex, "RegExp") ) {
         throw "fullRegex must be a RegExp object";
      }

      if( partialRegex != undefined && partialRegex != null && !isClassName(partialRegex, "RegExp") ) {
         throw "partialRegex must be a RegExp object";
      }

      destination[ id ] = { "fullRegex"    : fullRegex,
                            "partialRegex" : partialRegex };
   }

   $.fn.restrictedTextField = function( options ) {
      var settings = $.extend( {
         type                : null,
         preventInvalidInput : true,
         usePatternAttr      : false,
         logger              : undefined
      }, options );

      var EVENT_INPUT_IGNORED      = "inputIgnored",
          EVENT_VALIDATION_FAILED  = "validationFailed",
          EVENT_VALIDATION_SUCCESS = "validationSuccess";

      if( this.prop("tagName").toLowerCase() !== "input" ) {
         throw "RestrictedTextField can only be invoked on an input element";
      }

      if( settings.preventInvalidInput === true && settings.usePatternAttr === true ) {
         throw "Invalid configuration:  preventInvalidInput and usePatternAttr cannot both be true";
      }

      if( $.fn.restrictedTextField.types === undefined || $.fn.restrictedTextField.types === null ) {
         init();
      }

      function init() {
         var partialNegativeInt   = /^-\d*$/,

             partialPosNegFloat   = /^-?\d*\.?\d*$/,
             partialPositiveFloat = /^\d*\.?\d*$/,
             partialNegativeFloat = /^0*\.?$|^-\d*\.?\d*$/;

         if( $.fn.restrictedTextField.types === undefined || $.fn.restrictedTextField.types === null ) {
            $.fn.restrictedTextField.types = [];
         }

         var dest = $.fn.restrictedTextField.types = $.fn.restrictedTextField.types || [];

         _addType( dest, "alpha",                   /^[a-zA-Z]+$/            , null );
         _addType( dest, "upperAlpha",              /^[A-Z]+$/               , null );
         _addType( dest, "lowerAlpha",              /^[a-z]+$/               , null );
         _addType( dest, "alphaSpace",              /^[a-zA-Z ]+$/           , null );
         _addType( dest, "upperAlphaSpace",         /^[A-Z ]+$/              , null );
         _addType( dest, "lowerAlphaSpace",         /^[a-z ]+$/              , null );
         _addType( dest, "alphanumeric",            /^[a-zA-Z\d]+$/          , null );
         _addType( dest, "upperAlphanumeric",       /^[A-Z\d]+$/             , null );
         _addType( dest, "lowerAlphanumeric",       /^[a-z\d]+$/             , null );
         _addType( dest, "alphanumericSpace",       /^[a-zA-Z\d ]+$/         , null );
         _addType( dest, "upperAlphanumericSpace",  /^[A-Z\d ]+$/            , null );
         _addType( dest, "lowerAlphanumericSpace",  /^[a-z\d ]+$/            , null );
         _addType( dest, "int",                     /^-?\d+$/                , partialNegativeInt );
         _addType( dest, "positiveInt",             /^\d+$/                  , null );
         _addType( dest, "negativeInt",             /^0+$|^-\d+$/            , partialNegativeInt );
         _addType( dest, "strictInt",               /^0+$|^-?0*[1-9]\d*$/    , partialNegativeInt );
         _addType( dest, "strictPositiveInt",       /^\d+$/                  , null );
         _addType( dest, "strictNegativeInt",       /^0+$|^-0*[1-9]\d*$/     , partialNegativeInt );
         _addType( dest, "float",                   /^-?\d*\.?\d+$/          , partialPosNegFloat );
         _addType( dest, "positiveFloat",           /^\d*\.?\d+$/            , partialPositiveFloat );
         _addType( dest, "negativeFloat",           /^-?0+$|^-?0*\.?0+$|^-\d*\.?\d+$/ , partialNegativeFloat );
         _addType( dest, "strictFloat",             /^0*\.0+$|^-?0*\.\d*[1-9]$|^-?\d*\.\d*[1-9]\d*$|^-?0*[1-9]\d*?\d*\.\d+$/  , partialPosNegFloat );
         _addType( dest, "strictPositiveFloat",     /^0*\.0+$|^0*\.\d*[1-9]$|^\d*\.\d*[1-9]\d*$|^\d*[1-9]*\d*\.\d+$/     , partialPositiveFloat );
         _addType( dest, "strictNegativeFloat",     /^0*\.0+$|^-0*\.\d*[1-9]$|^-\d*\.\d*[1-9]\d*$|^-[^0]\d*?\d*\.\d+$/   , partialNegativeFloat );
         _addType( dest, "money",                   /^-?\d+\.\d{2}$/                   ,  /^-?\d*\.?\d{0,2}$/ );
         _addType( dest, "positiveMoney",           /^\d+\.\d{2}$/                     ,  /^\d*\.?\d{0,2}$/ );
         _addType( dest, "negativeMoney",           /^0\.00$|^-\d+\.\d{2}$/            ,  /^0*\.?0?$|^-\d*\.?\d{0,2}$/ );
         _addType( dest, "accountingMoney",         /^\d*\.?\d{2}$|^\(\d*\.?\d{2}\)$/  ,  /^\(?0*\.?0{0,2}$|^[\.\d]$|^\.$|^\d*\.\d{0,2}$|^\(\d*\.?$|^\(\d*\.\d{1,2}?\)?$/ );
         _addType( dest, "negativeAccountingMoney", /^0\.00$|^\(\d*\.?\d{2}\)$/        ,  /^\(?0*\.?0{0,2}$|^\(\d*\.?$|^\((\d*\.\d{1,2}?)\)?$/ );

         // Credit card regular expressions were written according to the formats described by https://en.wikipedia.org/wiki/Payment_card_number as of December 2016.
         // Note:  these formats evolve over time.

         // Prefix: 34, 37; Lengths: 15
         _addType( dest, "americanExpress",  /^3[47]\d{13}$/   ,  /^3[47]?$|^3[47]\d{0,13}$/ );

         // Prefix: 4; Lengths: 13, 16, 19
         _addType( dest, "visa",  /^4(\d{12}|\d{15}|\d{18})$/  ,  /^4\d{0,18}$/ );

         // Prefix: 51-55, 2221-2720; Lengths: 16
         _addType( dest,
                   "masterCard",
                   /^5[12345]\d{14}$|^(222[1-9]|22[3-9][0-9]|2[3-6][0-9]{2}|27[01][0-9]|2720)\d{12}$/  ,
                   joinRegex(
                              // Prefix 51-55, plus padding out to 16 digits
                              /^5([12345]?|[12345]\d{0,14})$/ ,

                              // Prefix 2221-2720 breakdown:

                                 // 22-27
                                 /^2[2-7]?$/ ,

                                 // 222-272
                                 /^(22[2-9]|2[3-6][0-9]|27[0-2])$/ ,

                                 // 2221-2720 plus padding out to 16 digits
                                 /^(222[1-9]|22[3-9][0-9]|2[3-6][0-9]{2}|27[01][0-9]|2720)\d{0,12}$/
                   )
         );

         // Prefix: 6011, 622126-622925, 644-649, 65; Lengths: 16, 19
         _addType( dest,
                   "discover",
                   /^6011(\d{12}|\d{15})$|^65(\d{14}|\d{17})$|^64[4-9](\d{13}|\d{16})$|^(62212[6-9]|6221[3-9][0-9]|622[2-8][0-9]{2}|6229[01][0-9]|62292[0-5])(\d{10}|\d{13})$/ ,
                   joinRegex(
                              // Prefix 6011
                              /^60?$|^601?$|^6011?$|^6011\d{0,15}$/ ,

                              // Prefix 622126-622925 breakdown:

                                 // 6 handled by the regex for the '65' prefix

                                 // 62, 622
                                 /^622?$/ ,

                                 // 6221-6229
                                 /^622[1-9]$/ ,

                                 // 62212-62292
                                 /^(6221[2-9]|622[2-8][0-9]|6229[0-2])$/ ,

                                 // 622126-622925 plus padding out to 16 or 19 digits
                                 /^(62212[6-9]|6221[3-9][0-9]|622[2-8][0-9]{2}|6229[01][0-9]|62292[0-5])(\d{0,10}|\d{0,13})$/ ,

                              // Prefix 644-649
                              /^64?$|^64[4-9]\d{0,16}$/ ,

                              // Prefix 65
                              /^6(5?|5\d{0,17})$/
                   )
         );

         // All credit card types
         _addType( dest, "creditCard", joinTypes(true, "americanExpress", "visa", "masterCard", "discover"), joinTypes(false, "americanExpress", "visa", "masterCard", "discover") );

         // Not as stringent as a formal credit card type - just performs Luhn validation.  Use this if you need Luhn validation for
         // a non-credit card number, or if you just don't like the idea of using the credit card types.  After all, you can just
         // leave it to your payment card processor to reject an invalid credit card number, while still using Luhn validation to
         // reject a number that has no possibility of being valid.  This is the safest option, but the choice is yours.
         _addType( dest, "luhnNumber",  /^\d+$/  ,  null );

         _addType( dest, "usZip",        /^\d{5}(-\d{1,4})?$/   , /^(\d{1,5}|\d{5}-|\d{5}-\d{1,4})$/ );
         _addType( dest, "usZip5",       /^\d{5}$/              , /^\d{1,5}$/ );
         _addType( dest, "usZipSuffix",  /^\d{1,4}$/            , null );
      }

      var regexes = $.fn.restrictedTextField.types[ settings.type ];

      if( isNothing(regexes) ) {
         if( $.fn.RestrictedTextFieldConfig.customTypes !== undefined && $.fn.RestrictedTextFieldConfig.customTypes !== null ) {
            regexes = $.fn.RestrictedTextFieldConfig.customTypes[ settings.type ];
         }

         if( isNothing(regexes) ) {
            throw "Invalid type: " + settings.type;
         }
      }

      /**
       * Combines multiple regular expression types into one
       *
       * @param [0] : boolean - true for fullRegex, false for partialRegex
       * @param vararg of regular expression IDs
       *
       * @return A single regular expression satisfying all specified types
       */
      function joinTypes() {
         if( arguments.length < 3 ) {
            throw "Minimum 3 arguments: boolean, id, id";
         }

         var isFullRegex = arguments[0];
         var src = $.fn.restrictedTextField.types;
         var regex = new RegExp( isFullRegex ? src[arguments[1]].fullRegex : src[arguments[1]].partialRegex );

         for( var i = 2; i < arguments.length; i++ ) {
            regex = joinRegex( regex, (isFullRegex ? src[arguments[i]].fullRegex : src[arguments[i]].partialRegex) );
         }

         return regex;
      }

      /**
       * Combines multiple regular expressions into one
       *
       * @param vararg of regular expression objects
       *
       * @return A single regular expression satisfying all provided regular expressions
       */
      function joinRegex() {
         if( arguments.length < 2 ) {
            throw "Minimum 2 RegExp arguments";
         }

         var regex = new RegExp( arguments[0] );

         for( var i = 1; i < arguments.length; i++ ) {
            regex = new RegExp( regex.source + "|" + arguments[i].source );
         }

         return regex;
      }

      return this.each( function() {
         var jqThis = $( this );
         var valueBeforeCommit = "";

         if( settings.usePatternAttr === true ) {
            var str = regexes.fullRegex.toString();
            jqThis.prop( "pattern", str[0] === "/" && str[str.length - 1] === "/" ? str.substring(1, str.length - 1) : str );
         } else {
            jqThis.on( "keydown paste", function(event) {
               valueBeforeCommit = this.value;
            } );

            jqThis.on( "input", function(event) {
               log( "caught input - new value: " + this.value );

               var jqThis = $( this );

               var passesPartialRegex = false,
                   passesFullRegex    = false;

               var responseEvent = settings.preventInvalidInput ? null : EVENT_VALIDATION_FAILED;

               if( this.value.length === 0 ) {
                  passesPartialRegex = true;
                  passesFullRegex = true;

                  if( !settings.preventInvalidInput ) {
                     responseEvent = EVENT_VALIDATION_SUCCESS;
                  }
               } else {
                  passesFullRegex = regexes.fullRegex.test( this.value );

                  if( regexes.partialRegex !== undefined && regexes.partialRegex !== null ) {
                     passesPartialRegex = regexes.partialRegex.test( this.value );
                  }

                  if( passesFullRegex ) {
                     log( "passes full regex" );

                     if( !settings.preventInvalidInput ) {
                        responseEvent = EVENT_VALIDATION_SUCCESS;
                     }
                  } else {
                     log( "fails full regex" );

                     if( passesPartialRegex ) {
                        log( "passes partial regex" );

                        if( !settings.preventInvalidInput ) {
                           responseEvent = EVENT_VALIDATION_SUCCESS;
                        }
                     } else {
                        log( "fails partial regex" );

                        if( !settings.preventInvalidInput ) {
                           responseEvent = EVENT_VALIDATION_FAILED;
                        }
                     }
                  }

                  if( !passesPartialRegex && !passesFullRegex ) {
                     if( settings.preventInvalidInput ) {
                        responseEvent = EVENT_INPUT_IGNORED;
                        jqThis.val( valueBeforeCommit );
                        log( "reverted field to \"" + valueBeforeCommit + "\"" );
                     }
                  }
               }

               if( responseEvent != null ) {
                  log( "triggering " + responseEvent + " event " );
                  jqThis.trigger( responseEvent );
               }
            } );
         }

         jqThis.on( "blur", function() {
            var type = settings.type;
            var passesFullRegex = false;
            var responseEvent = EVENT_VALIDATION_FAILED;

            if( ((type === "money" || type === "positiveMoney" || type === "negativeMoney") && !isNaN(this.value)) ||
                ((type === "accountingMoney" || type ==="negativeAccountingMoney") && !isNaN(this.value.replace(/^\((.*)\)$/, "$1"))) ) {
               var formatted = formatMoney( this.value );

               passesFullRegex = regexes.fullRegex.test( formatted );

               if( passesFullRegex ) {
                  this.value = formatted;
                  log( "blur:  formatted " + settings.type + " field to " + formatted );
               }
            } else if( type === "americanExpress" || type === "visa" || type === "masterCard" || type === "discover" || type === "creditCard" || type === "luhnNumber" ) {
               if( settings.usePatternAttr === false ) {
                  passesFullRegex = regexes.fullRegex.test( this.value );

                  if( passesFullRegex ) {
                     passesFullRegex = luhnCheck( this.value );
                     log( passesFullRegex ? "passes Luhn check" : "fails Luhn check" );
                  }
               }
            } else if( settings.usePatternAttr === false ) {
               passesFullRegex = regexes.fullRegex.test( this.value );
            }

            if( settings.usePatternAttr === false ) {
               if( this.value.length === 0 || passesFullRegex ) {
                  responseEvent = EVENT_VALIDATION_SUCCESS;
               }

               log( "triggering " + responseEvent + " event " );

               jqThis.trigger( responseEvent );
            }
         } );

         function formatMoney( value ) {
            if( isNothing(value) ) return value;

            var formatted = value;

            value = trimLeadingZero( value );

            var len         = value.length;
            var sign        = value[0] === "-" ? "-" : "";
            var openParen   = value[0] === "(" ? "(" : "";
            var closeParen  = value[ len - 1 ] === ")" ? ")" : "";
            var decimalIdx  = value.indexOf( "." );
            var integerPart = value.substring( openParen === "" && sign === "" ? 0 : 1,
                                               decimalIdx > 0 ? decimalIdx -1 : (closeParen === ")" ? len - 1 : len) );
            var decimalPart = "00";

            if( decimalIdx > -1 ) {
               integerPart = value.substring( openParen === "" && sign === "" ? 0 : 1, decimalIdx );
               decimalPart = value.substring( decimalIdx + 1, closeParen === "" ? len : len - 1 );
            }

            if( (integerPart === "" || parseInt(integerPart) === 0) && (parseInt(decimalPart) === 0) ) {
               formatted = "0.00";
            } else {
               formatted = openParen
                         + sign
                         + (integerPart === "" ? "0" : integerPart)
                         + "."
                         + decimalPart
                         + (decimalPart.length === 1 ? "0" : "")
                         + closeParen;
            }

            return formatted;

            function trimLeadingZero( str ) {
               return str.replace( /^(-?\(?)0*(\d+\.\d+)/, "$1$2" );
            }
         }
      } );

      function luhnCheck( numStr ) {
         var luhnNums = [];
         var sum = 0;
         var checkDigit = 0;

         if( isNaN(numStr) ) {
            throw "Value [" + numStr + "] is not numeric";
         }

         var doubleMe = true;

         for( var i = numStr.length - 2; i >= 0; i-- ) {
            var num = parseInt( numStr[i] );

            if( doubleMe ) {
               var x2 = num * 2;
               luhnNums[i] = x2 > 9 ? x2 - 9 : x2;
            } else {
               luhnNums[i] = num;
            }

            sum += luhnNums[i];
            doubleMe = !doubleMe;
         }

         checkDigit = (sum * 9) % 10;
         sum += checkDigit;

         return sum % 10 === 0 && parseInt( numStr[numStr.length - 1] ) === checkDigit;
      }

      function log( msg ) {
         if( settings.logger && typeof settings.logger === "function" ) {
            settings.logger( "jquery.restrictedtextfield.js:  " + msg );
         }
      }
   };

   function isNothing( value ) {
      return value === undefined || value === null || value.length < 1;
   }

   function isClassName( obj, className ) {
      var name = Object.prototype.toString.call( obj ).slice( 8, -1 );
      return obj !== undefined && obj !== null && name.toUpperCase() === className.toUpperCase();
   }
}( jQuery ));
